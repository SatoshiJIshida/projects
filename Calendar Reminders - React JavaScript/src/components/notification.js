import React from "react";
import "../index.css";

/**
 * Notification Component.
 * @author  [Satoshi Ishida]
 *
 * @param   {boolean} show          - true or false to show the modal
 * @param   {object}  children      - to show child elements
 * @returns {string}  showHideClass - CSS classes chosen based on boolean value to show or hide the modal
 * @returns {object}  children      - child elements in modal
 */
const Notification = ({ show, children }) => {
  const showHideClass = show
    ? "animateUp modal display-block"
    : "modal display-none";
  return (
    <div className={showHideClass}>
      <section>{children}</section>
    </div>
  );
};

// For single month view.
const Notification2 = ({ show, children }) => {
  const showHideClass = show
    ? "animateUp modal2 display-block"
    : "modal2 display-none";
  return (
    <div className={showHideClass}>
      <section>{children}</section>
    </div>
  );
};

export { Notification, Notification2 };
