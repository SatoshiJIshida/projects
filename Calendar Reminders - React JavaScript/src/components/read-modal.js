import React from "react";
import "../index.css";

/**
 * Read Modal Component.
 * @author  [Satoshi Ishida]
 *
 * @param   {boolean} show          - true or false to show the modal
 * @param   {object}  children      - to show child elements
 * @returns {string}  showHideClass - CSS classes chosen based on boolean value to show or hide the modal
 * @returns {object}  children      - child elements in modal
 */
const ReadModal = ({ show, children }) => {
  const showHideClass = show ? "modal display-block" : "modal display-none";
  return (
    <div className={showHideClass}>
      <section>{children}</section>
    </div>
  );
};

// For single month view.
const ReadModal2 = ({ show, children }) => {
  const showHideClass = show ? "modal2 display-block" : "modal2 display-none";
  return (
    <div className={showHideClass}>
      <section>{children}</section>
    </div>
  );
};

export { ReadModal, ReadModal2 };
