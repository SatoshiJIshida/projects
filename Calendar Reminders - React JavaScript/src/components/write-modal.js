import React from "react";
import "../index.css";

/**
 * Write Modal Component.
 * @author  [Satoshi Ishida]
 *
 * @param   {boolean} show          - true or false to show the modal
 * @param   {object}  children      - to show child elements
 * @returns {string}  showHideClass - CSS classes chosen based on boolean value to show or hide the modal
 * @returns {object}  children      - child elements in modal
 */
const WriteModal = ({ show, children }) => {
  const showHideClass = show
    ? "animateUp modal display-block"
    : "modal display-none";
  return (
    <div className={showHideClass}>
      <section>{children}</section>
    </div>
  );
};

// For single month view.
const WriteModal2 = ({ show, children }) => {
  const showHideClass = show
    ? "animateUp modal2 display-block"
    : "modal2 display-none";
  return (
    <div className={showHideClass}>
      <section>{children}</section>
    </div>
  );
};

export { WriteModal, WriteModal2 };
