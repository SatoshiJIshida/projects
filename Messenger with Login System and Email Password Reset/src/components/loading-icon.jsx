import React from "react";
import "../index.css";

/**
 * LoadingIcon functional component.
 * @returns {component} LoadingIcon component.
 * @author [Satoshi Ishida]
 */
export const LoadingIcon = () => {
  return (
    <div key="loadingIcon" className="loading-spin-container">
      <div className="loading-spin"></div>
    </div>
  );
};
