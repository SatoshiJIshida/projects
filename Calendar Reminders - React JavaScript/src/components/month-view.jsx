import React from "react";
import firebase from "firebase/app";
import "../index.css";
import months from "../custom/months";
import { WriteModal2 } from "./write-modal";
import { ReadModal2 } from "./read-modal";
import { Notification2 } from "./notification";
import { date, fullDate, year } from "../custom/date-pattern";
import { monthSelected, savedDays, title, dbDate } from "./year-view";
import { enterYearView } from "../helpers/enter-year-view";

const currentDate = date(), // e.g. January1.
      curFullDate = fullDate(); // e.g. Monday 1 January for titles.

var isTurnedOn = false, // show notification.
    userSelect = "", // current date user selected.
    title2 = title, // full date title for specific selected date, not system date.
    dbDate2 = dbDate; // reminder date e.g. January1 used in logic to get data from database.

/**
 * Month View Component.
 * @author [Satoshi Ishida]
 */
export default class MonthView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      months: months, // get months data.
      month: null, // selected month index.
      selected: null, // selected date.
      isOpen: false, // show/hide reminder input modal.
      isShowReminder: false, // show/hide reminder.
      isTurnedOff: false, // close notification.
      dateItem: [], // dateItem e.g. January1.
      value: [], // reminder text from text area.
      getReminder: [], // get database data.
    };
  }

  /**
   * dynamicClass -> colour change for reminders being saved.
   * @param   {object}  month        - Month of the year
   * @param   {string}  j            - Date in a month e.g. 1-31
   * @returns {string}  dynamicClass - CSS className
   */
  dynamicClass = (month, j) => {
    let dynamicClass = "";
    // imported savedDays array keeping track of saved reminder dates.
    if (savedDays.includes(month.title + j)) { // if it includes e.g. January1.
      dynamicClass = "colourBtn2"; // colour change for saved reminder.
      return dynamicClass;
    } else {
      dynamicClass = "dateBtn2"; // default.
      return dynamicClass;
    }
  };

  /**
   * disabledClass -> color change for days not in use in each month.
   * @param   {object} disabledDays1 - days not in use in row1
   * @param   {object} disabledDays2 - days not in use in final rows
   * @returns {string} dynamicClass  - CSS className
   */
  disabledClass = (disabledDays1, disabledDays2) => {
    let dynamicClass = "";

    if (disabledDays1 || disabledDays2) {
      dynamicClass = "disabled2"; // stop interactivity.
      return dynamicClass;
    } else {
      dynamicClass = "dateBtn2"; // default.
      return dynamicClass;
    }
  };

  /**
   * selectedDate -> date in a month that user selects.
   * @param {object} month - Month in a year
   * @param {string} j     - Date in a month e.g. 1-31
   */
  selectedDate = (month, j) => {
    this.setState({ month: month, selected: j, isOpen: true }); // set month and date selected and input reminder modal opens.
    userSelect = j; // also store selected date here to use elsewhere.

    if (savedDays.includes(month.title + j)) { // if we already have a reminder for the selected date.
      this.setState({ isOpen: false, isShowReminder: true }); // instead show the reminder modal.
      this.makeTitles(month, j); // make title to use for the reminder selected date.
      this.getReminder(month.title + j); // send date reference parameter to getReminder().
    } else {
      return;
    }
  };

  /**
   * function to make day and date titles to concatenate into reminder title e.g. Monday 1 January.
   * @param {object} month - Month in a year
   * @param {string} j     - Date in a month e.g. 1-31
   */
  makeTitles = (month, j) => {
    let makeDate = "";
    /**
     * Getting data ->
     * index0 is the day title e.g. Monday.
     * index1 is the array of dates that occur on a Monday that selected month.
     */
    if (month.monday[1].includes(userSelect)) {
      // if month's Monday dates array includes current selected date.
      makeDate = month.monday[0] + " " + j + " " + month.title; // extract the Monday title to concatenate with j and month title -> to produce e.g. Monday 1 January.
      title2 = makeDate; // store the selected title to use.
    } else if (month.tuesday[1].includes(userSelect)) {
      makeDate = month.tuesday[0] + " " + j + " " + month.title;
      title2 = makeDate;
    } else if (month.wednesday[1].includes(userSelect)) {
      makeDate = month.wednesday[0] + " " + j + " " + month.title;
      title2 = makeDate;
    } else if (month.thursday[1].includes(userSelect)) {
      makeDate = month.thursday[0] + " " + j + " " + month.title;
      title2 = makeDate;
    } else if (month.friday[1].includes(userSelect)) {
      makeDate = month.friday[0] + " " + j + " " + month.title;
      title2 = makeDate;
    } else if (month.saturday[1].includes(userSelect)) {
      makeDate = month.saturday[0] + " " + j + " " + month.title;
      title2 = makeDate;
    } else if (month.sunday[1].includes(userSelect)) {
      makeDate = month.sunday[0] + " " + j + " " + month.title;
      title2 = makeDate;
    } else {
      return;
    }
  };

  /**
   * setupReminder -> for notifications, get the data from database calling getReminder();
   */
  setupReminder = () => {
    if (savedDays.includes(currentDate)) {
      title2 = curFullDate; // make full title for it using system date.
      this.setState({
        isShowReminder: true, // show the reminder modal.
        isTurnedOff: true, // hide the notification modal.
        dateItem: currentDate, // set dateItem as currentDate pattern taken from system date (not user selected date).
      });
      this.getReminder(currentDate); // get the reminder with the current system date.
    } else {
      return;
    }
  };

  /**
   * getReminder -> get reminder text data from database.
   * @param {string} dateItem - month.title + j e.g. January1
   */
  getReminder = (dateItem) => {
    const reminder = firebase.database().ref("reminders/" + dateItem + "/text");
    let err;
    reminder.on("value", (snapshot) => {
      const data = snapshot.val();
      this.extractReminder(err, data);
    });
  };

  /**
   * extractReminder -> after getting the reminder data object, we need to store it.
   * @param {object} err  - empty variable to use for error catch
   * @param {object} data - reminder data from database
   */
  extractReminder = (err, data) => {
    new Promise((resolve, reject) => {
      if (err) {
        reject(err);
      } else {
        resolve(data);
      }
    })
      .then((data) => {
        this.setState({ getReminder: data }); // set the data in getReminder state.
      })
      .catch((err) => {
        throw err; // otherwise, there was an error and terminate.
      });
  };

  /**
   * displayReminder -> setting up the reminder modal.
   * @returns {object} title2 + year()         - day + date + month + year e.g. Monday 1 January 2021
   * @returns {object} this.state.getReminder  - reminder text
   * @returns {object} button                  - delete reminder button
   * @returns {object} button                  - close the reminder modal
   */
  displayReminder = () => {
    return (
      <div>
        <div className="reminderTitle2">{title2 + " " + year()}</div>
        <div className="reminderText2">{this.state.getReminder}</div>
        <button
          className="deleteReminder2"
          onClick={() => this.removeReminder()}
        >
          Delete Reminder
        </button>
        <button
          className="reminderClose2"
          onClick={() => this.setState({ isShowReminder: false })}
        >
          &times;
        </button>
      </div>
    );
  };

  /**
   * removeReminder -> remove the reminder from the savedDays array and the database.
   */
  removeReminder = () => {
    if (this.state.month) { // if month state exists because user clicked a date.
      dbDate2 = this.state.month.title + this.state.selected; // e.g. if you clicked on January1 then that is what will be deleted.
      const i = savedDays.indexOf(dbDate2);
      savedDays.splice(i, 1); // remove 1 element from index where our selected date is.
      firebase // also remove from firebase database.
        .database()
        .ref("reminders/" + dbDate2)
        .remove(); // remove reminder.
      firebase
        .database()
        .ref("dateItem/" + dbDate2)
        .remove(); // remove reminder date.
      this.setState({ isShowReminder: false }); // close reminder after removing it.
    } else {
      dbDate2 = this.state.dateItem; // else user didn't click a date and we remove a loaded notification reminder -> in setupReminder() we set it as currentDate.
      const i = savedDays.indexOf(dbDate2);
      savedDays.splice(i, 1); // remove 1 element from index where our selected date is.
      firebase // also remove from firebase database.
        .database()
        .ref("reminders/" + dbDate2)
        .remove(); // remove reminder.
      firebase
        .database()
        .ref("dateItem/" + dbDate2)
        .remove(); // remove reminder date.
      this.setState({ isShowReminder: false }); // close reminder after removing it.
    }
  };

  /**
   * displayInputModal -> setup the input reminder modal.
   * @returns {string} title
   * @returns {object} text area - input reminder text area
   * @returns {object} button    - submit reminder button
   * @returns {object} button    - close input reminder modal
   */
  displayInputModal = () => {
    if (this.state.isOpen === false) {
      return null;
    } else {
      return (
        <div>
          <form key={this.state.selected} className="form">
            <div className="reminderTitle2">Enter Reminder:</div>
            <textarea
              className="textArea2"
              value={this.state.value}
              onChange={this.handleChange}
            />
            <button className="submit" onClick={this.handleSubmit}>
              Submit
            </button>
          </form>
          <button
            className="inputClose"
            onClick={() => this.setState({ isOpen: false })}
          >
            &times;
          </button>
        </div>
      );
    }
  };

  /**
   * handleChange -> reminder text input.
   * @param {event} event
   */
  handleChange = (event) => {
    this.setState({ value: event.target.value }); // reminder text stored in 'value' state.
  };

  /**
   * handleSubmit -> submission of reminders.
   * @param {event} event
   */
  handleSubmit = (event) => {
    event.preventDefault();
    const dateItem = this.state.month.title + this.state.selected; // e.g. January1.

    savedDays.push(dateItem); // push date of reminder to savedDays array.
    this.writeUserData(dateItem); // and to firebase database.
    this.setState({ isOpen: false, value: "" }); // After submission, hide input reminder modal and clear text area.
  };

  /**
   * writeUserData -> write reminder to database.
   * @param {string} dateItem - this.state.month.title + this.state.selected e.g. January1
   */
  writeUserData = (dateItem) => {
    firebase
      .database()
      .ref("dateItem/" + dateItem)
      .set({
        dateItem, // save dateItem.
      });
    firebase
      .database()
      .ref("reminders/" + dateItem)
      .set({
        text: this.state.value, // save that dateItem's text reminder value.
      });
  };

  /**
   * notification -> check if current date is stored and open the notification.
   */
  notification = () => {
    /**
     * You need to also check that isTurnedOff (notification checker) is false, otherwise the notification will always stay on screen
     * because notification() is in the render and will constantly check for a reminder on the current day.
     * This is also why isTurnedOn is a boolean global variable, because you can't constantly call to setState from render().
     */
    if (savedDays.includes(currentDate) && this.state.isTurnedOff === false) {
      isTurnedOn = true; // open notification modal if today has a reminder.
    } else {
      isTurnedOn = false; // else there is no reminder today.
    }
  };

  /**
   * displayNotification -> modal for notification, if current day has a reminder.
   * @returns {object} string + curFullDate + year() - title of notification
   * @returns {object} button                        - view reminder button calls this.setupReminder()
   * @returns {object} button                        - to close the notification
   */
  displayNotification = () => {
    return (
      <div>
        <div className="reminderTitle2">
          {"You have a reminder: " + curFullDate + " " + year()}
        </div>
        <button className="viewReminder" onClick={() => this.setupReminder()}>
          View Reminder
        </button>
        <button
          className="notifClose"
          onClick={() => this.setState({ isTurnedOff: true })}
        >
          &times;
        </button>
      </div>
    );
  };

  /**
   * Class Render.
   * @returns {object} button                - month title with year e.g. January 2021
   * @returns {string} dayLetter             - day letters for the date numbers in each month
   * @returns {object} button                - disabledDays1 buttons (greyed out)
   * @returns {object} button                - days buttons (dates in use)
   * @returns {object} button                - disabledDays2 buttons (greyed out)
   * @returns {object} button                - go back to year view
   * @returns {function} this.notification() - check to display notification
   * @returns {function} WriteModal          - reminder input modal
   * @returns {function} ReadModal           - reminder read modal
   * @returns {function} Notification        - notification modal
   */
  render() {
    return this.state.months.outputMonth.map((i) => { // use index of outputMonth array that contains month objects.
      const month = this.state.months.month[i], // get data from /custom/months.js
            singleMonth = [];

      let dayLetters = [],
          days = [],
          disabledDays1 = [],
          disabledDays2 = [];

      if (month.id === monthSelected) { // if month[i] id is the same as month[i] id selected in Year View.
        singleMonth.push(month); // push that selected month the user clicked in the Year View to be rendered here as a single Month View.
        dayLetters = singleMonth[0].dayLetters;
        days = singleMonth[0].days;
        disabledDays1 = singleMonth[0].disabledDays1;
        disabledDays2 = singleMonth[0].disabledDays2;

        return [
          <div>
            <div className="titleContainer2">
              <button className={"monthTitle2"}>
                {month.title + " " + year()}
              </button>
            </div>

            <div className="singleMonthContainer">
              <div className="dayLetterTwo1">{dayLetters[0]}</div>
              <div className="dayLetterTwo2">{dayLetters[1]}</div>
              <div className="dayLetterTwo3">{dayLetters[2]}</div>
              <div className="dayLetterTwo4">{dayLetters[3]}</div>
              <div className="dayLetterTwo5">{dayLetters[4]}</div>
              <div className="dayLetterTwo6">{dayLetters[5]}</div>
              <div className="dayLetterTwo7">{dayLetters[6]}</div>

              <div className="daysContainer2">
                {disabledDays1.map((j) => {
                  return (
                    <button
                      key={j}
                      className={this.disabledClass(disabledDays1)}
                    >
                      {j}
                    </button>
                  );
                })}

                {days.map((j) => {
                  return (
                    <button
                      key={j}
                      className={this.dynamicClass(month, j)}
                      onClick={() => this.selectedDate(month, j)}
                    >
                      {j}
                    </button>
                  );
                })}

                {disabledDays2.map((j) => {
                  return (
                    <button
                      key={j}
                      className={this.disabledClass(disabledDays2)}
                    >
                      {j}
                    </button>
                  );
                })}
              </div>
              <button className={"retBtn2"} onClick={() => enterYearView()}>
                Return
              </button>
            </div>
          </div>,

          this.notification(),
          <WriteModal2 show={this.state.isOpen}>
            {this.displayInputModal()}
          </WriteModal2>,
          <ReadModal2 show={this.state.isShowReminder}>
            {this.displayReminder()}
          </ReadModal2>,
          <Notification2 show={isTurnedOn}>
            {this.displayNotification()}
          </Notification2>,
        ];
      }
      return singleMonth;
    });
  }
}

export { title2, dbDate2 };
