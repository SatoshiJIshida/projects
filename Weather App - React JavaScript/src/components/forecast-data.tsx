import React, { useState } from "react";
import "../index.css";
import { locate } from "./today-data";
import { day } from "../helpers/day";

/**
 * Weather forecast for 5 days after today.
 * @author [Satoshi Ishida]
 */
const Forecast = () => {
  const [weather, setWeather] = useState(Object);
  const [check, setCheck] = useState(false); // check weather data was fetched.
  const [clicked, setClicked] = useState(false); // check btn was clicked.

  /**
   * getWeather from open weather map API.
   * @returns {object} result - JSON object of weather details or error
   */
  async function getWeather() {
    try {
      fetch(
        `https://api.openweathermap.org/data/2.5/forecast?q=${locate}&appid=9d5bdc3539df34107238c1d54d0712b0&units=metric&cnt=5`
      )
        .then((result: any) => {
          if (result.ok) {
            console.log("Success: " + result.status); // 200 for success.
            return result.json();
          } else {
            console.log("Failure: " + result.status); // will be a different error code.
            return result.status;
          }
        })
        .then((object: any) => {
          setWeather(object); // put this json data into weather state.
          setCheck(true); // set true because weather data was successfully fetched.
        });
    } catch (err) {
      console.log(err);
      return err;
    }
  }

  /**
   * getIcon from open weather map API.
   * @returns {object} img - weather icon
   */
  function getIcon() {
    const data: Array<Object> = [];

    if (check === true) { // we check if data was fetched.
      for (let i: number = 0; i < weather.list.length; i++) { // this needs to go here because outside of check, it might not exist yet.
        data.push(weather.list[i]); // equals gives you a combined array of the data, push gives you separate arrays for each day's data.
      }
      return data.map((i: any) => {
        const icon: Array<Object> = [];
        icon.push(i?.weather[0]?.icon); // get the weather icon image for day index.
        const link: string = "http://openweathermap.org/img/w/" + icon + ".png";
        return (
          <div className="forecastIcon">
            <img src={link} alt="weather icon" />
          </div>
        );
      });
    }
  }

  /**
   * getTemp from JSON data in weather state.
   * @returns {string} temp - today's temp in degrees celsius.
   */
  function getTemp() {
    const data: Array<Object> = [];

    if (check === true) {
      for (let i: number = 0; i < weather.list.length; i++) { // this needs to go here because outside of check, it might not exist yet.
        data.push(weather.list[i]);
      }
      return data.map((i: any) => {
        const temp: Array<Object> = [];
        temp.push(i?.main?.temp);
        return (
          <div className="forecastTemp">
            <p>{temp}°C</p>
          </div>
        );
      });
    }
  }

  /**
   * getWeatherInfo -> get remaining data that I want to use from the JSON data in the weather state.
   * @returns {string} main     - description of weather for today.
   * @returns {string} high     - max temp for today.
   * @returns {string} low      - min temp for today.
   * @returns {string} humidity - humidity for today.
   */
  function getWeatherInfo() {
    const data: Array<Object> = [];

    if (check === true) {
      for (let i: number = 0; i < weather.list.length; i++) { // this needs to go here because outside of check, it might not exist yet.
        data.push(weather.list[i]);
      }
      return data.map((i: any) => {
        const high: Array<Object> = i?.main?.temp_max;
        const low: Array<Object> = i?.main?.temp_min;
        const humidity: Array<Object> = i?.main?.humidity;

        const main: string = weather.list[i]?.weather[0]?.main; // get the weather type data.
        return (
          <div className="forecastData">
            <p>{main}</p>
            <p>High: {high}°C</p>
            <p>Low: {low}°C</p>
            <p>Humidity: {humidity}%</p>
          </div>
        );
      });
    }
  }

  function onClick() {
    setClicked(true);
    getWeather();
  }

  function hide() {
    const show: string = "btnForecast";
    const hide: string = "hide";
    if (clicked === true) {
      return hide;
    } else {
      return show;
    }
  }

  const getDays: Array<string> = [];

  for (let i: number = 1; i < day().length; i++) {
    // we get the next 5 days for the forecast.
    getDays.push(day()[i]);
  }

  return (
    <React.Fragment>
      {/* Development button */}
      <button className={hide()} onClick={onClick}>
        Get 5 Day Forecast
      </button>

      <div className="forecastContainer">
        <div>
          {getDays.map((i: string) => {
            return (
              <div key={i} className="daysContainer">
                {clicked === true ? <p className="days">{i}</p> : null}
              </div>
            );
          })}
        </div>

        <div>
          {/* Icon */}
          {clicked === true ? getIcon() : null}
        </div>

        <div>
          {/* Weather Info */}
          {clicked === true ? getWeatherInfo() : null}
        </div>

        <div>
          {/* Temp Info */}
          {clicked === true ? getTemp() : null}
        </div>
      </div>
    </React.Fragment>
  );
};

export { Forecast };
