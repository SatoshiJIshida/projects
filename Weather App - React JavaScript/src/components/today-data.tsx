import React, { useState, useEffect } from "react";
import "../index.css";
import { day } from "../helpers/day";
import { Forecast } from "./forecast-data";

var locate: string = "";

/**
 * Today location and weather.
 * @author [Satoshi Ishida]
 */
const Today = () => {
  const xhr: XMLHttpRequest = new XMLHttpRequest();

  const [location, setLocation] = useState("Locating...");
  const [country, setCountry] = useState("");
  const [weather, setWeather] = useState(Object);
  const [check, setCheck] = useState(false); // check weather data was fetched.
  const [clicked, setClicked] = useState(false); // check btn was clicked.

  /**
   * Parent function getCoordinates() ->
   * If geolocation is successful for user, then calls success(), which calls getLocation()
   * else it calls error().
   */
  function getCoordinates() {
    const options: Object = {
      enableHighAccuracy: true,
      timeout: 5000, // the amount of time before the error callback is invoked.
      maximumAge: 0, // maximum cached position age.
    };

    /**
     * If successful in getting user location.
     * @param {object} pos - the user's position data
     */
    function success(pos: any) {
      const crd: any = pos.coords;
      const lat: number = crd.latitude;
      const long: number = crd.longitude;
      const coordinates: Array<number> = [lat, long];
      console.log(`Latitude: ${lat}, Longitude: ${long}`);
      getLocation(coordinates);
      return;
    }

    /**
     * If unsuccessful in getting the user location.
     * @param err - error status
     */
    function error(err: any) {
      console.warn(`ERROR(${err.code}); ${err.message}`);
      setLocation("Unable to retrieve your location");
      locate = location;
    }

    navigator.geolocation.getCurrentPosition(success, error, options);
  }

  /**
   * getLocation -> fetches location data.
   * @param {object} coordinates - user location coordinates
   */
  function getLocation(coordinates: Array<number>) {
    const lat: number = coordinates[0]; // lat horizontal.
    const long: number = coordinates[1]; // long vertical.
    // Using LocationIQ.
    xhr.open(
      // open() initializes a request.
      "GET",
      "https://us1.locationiq.com/v1/reverse.php?key=pk.02bc6b0c0ff276b8b043dc1804f6b1c0&lat=" +
        lat +
        "&lon=" +
        long +
        "&format=json",
      true
    );
    xhr.send(); // send() sends the request.
    xhr.onreadystatechange = processRequest; // onready is event handler, calls processRequest().
    xhr.addEventListener("readystatechange", processRequest, false); // also, the standard events listener (type, listener, option (false so handler is called on bubbling))
  }

  /**
   * processRequest is called from getLocation() event listener.
   * This is the listener function that runs on event.
   */
  function processRequest() {
    if (xhr.readyState === 4 && xhr.status === 200) { // 4 means DONE and 200 means Successful response.
      const response: any = JSON.parse(xhr.responseText); // get the response text.
      const town: string = response.address.town;
      const city: string = response.address.city;
      const country: string = response.address.country;

      if (country === "United Kingdom") {
        setCountry("GB");
      } else {
        setCountry(country);
      }

      if (town) {
        setLocation(town);
        locate = location; // so that we can export this into the weekly forecast.
      } else if (!town && city) {
        setLocation(city); // if user is in a city and not a town.
        locate = location;
      } else {
        return;
      }
    }
  }

  /**
   * getWeather from open weather map API.
   * @returns {object} result - JSON object of weather details or error
   */
  async function getWeather() {
    try {
      fetch(
        `https://api.openweathermap.org/data/2.5/weather?q=${location}&appid=9d5bdc3539df34107238c1d54d0712b0&units=metric`
      )
        .then((result: any) => {
          if (result.ok) {
            console.log("Success: " + result.status); // 200 for success.
            return result.json();
          } else {
            console.log("Failure: " + result.status); // will be a different error code.
            return result.status;
          }
        })
        .then((object: Object) => {
          setWeather(object); // put this json data into weather state.
          setCheck(true);
        });
    } catch (err) {
      console.log(err);
      return err;
    }
  }

  /**
   * getIcon from open weather map API.
   * @returns {object} img - weather icon
   */
  function getIcon() {
    // ?. is the optional chaining operator to read the value of the property located deep within a chain
    // of connected objects without having to check that each reference in the chain is valid.
    const icon: string = weather?.weather[0]?.icon; // get the weather icon image.
    const link: string = "http://openweathermap.org/img/w/" + icon + ".png"; // we can use the icon id in the url to get the image.

    return (
      <div id="icon">
        <img src={link} alt="weather icon" />
      </div>
    );
  }

  /**
   * getTemp from JSON data in weather state.
   * @returns {string} temp - today's temp in degrees celsius.
   */
  function getTemp() {
    return (
      <div>
        <p>{weather?.main?.temp}°C</p>
      </div>
    );
  }

  /**
   * getWeatherInfo -> get remaining data that I want to use from the JSON data in the weather state.
   * @returns {string} main     - description of weather for today.
   * @returns {string} temp_max - max temp for today.
   * @returns {string} temp_min - min temp for today.
   * @returns {string} humidity - humidity for today.
   */
  function getWeatherInfo() {
    const main: string = weather?.weather[0]?.main; // get the weather type data.
    return (
      <div>
        <p>{main}</p>
        <p>High: {weather?.main?.temp_max}°C</p>
        <p>Low: {weather?.main?.temp_min}°C</p>
        <p>Humidity: {weather?.main?.humidity}%</p>
      </div>
    );
  }

  function onClick() {
    if (location != "Locating..." && "Unable to retrieve your location") {
      setClicked(true);
      getWeather();
    } else {
      return;
    }
  }

  function hide() {
    const show: string = "btnDay";
    const hide: string = "hide";
    if (clicked === true) {
      return hide;
    } else {
      return show;
    }
  }

  useEffect(() => {
    getCoordinates(); // get location on mount.
  });

  return (
    <React.Fragment>
      <div className="dayContainer">
        <div className="dayData">
          {/* Title */}
          <h1>
            Today: {day()[0]} | {location} {country}
          </h1>

          {/* Development button */}
          <button className={hide()} onClick={onClick}>
            Get Weather
          </button>

          {/* Weather Info */}
          {check === true ? getWeatherInfo() : null}
        </div>

        <div className="dayIcon">
          {/* Icon */}
          {check === true ? getIcon() : null}
        </div>

        <h1 className="dayTemp">
          {/* Temp Info */}
          {check === true ? getTemp() : null}
        </h1>

        <div>{check === true ? <Forecast /> : null}</div>
      </div>
    </React.Fragment>
  );
};

export { Today, locate };
