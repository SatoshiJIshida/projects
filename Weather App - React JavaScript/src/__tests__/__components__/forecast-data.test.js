import { shallow } from "enzyme";
import { Forecast } from "../../components/forecast-data";

describe("Forecast Data Component", () => {
  it("should be defined", () => {
    expect(Forecast).toBeDefined();
  });

  it("should render correctly", () => {
    const newForecast = shallow(<Forecast name="forecast-test" />);
    expect(newForecast).toMatchSnapshot();
  });
});
