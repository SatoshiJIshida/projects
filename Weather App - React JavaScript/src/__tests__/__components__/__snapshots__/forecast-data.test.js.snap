// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast Data Component should render correctly 1`] = `ShallowWrapper {}`;
